Decks/Source/Deck.cpp

```diff
@@ -1,24 +1,67 @@
 #include "Deck.hpp"
 
+Deck::Deck()
+{
+  cards = std::make_shared<std::vector<Card>>(_cards);
+}
+
+uint Deck::findCard(Card& cardToFind)
+{
+  for(uint i = 0; i < this->cards->size(); ++i)
+  {
+    if(this->cards->at(i) == cardToFind)
+    {
+      return i;
+    }
+  }
+  throw std::runtime_error("Card not found");;
+}
+
 void Deck::addCard(Card cardToAdd)
 {
-  _cards.push_back(cardToAdd);
+  cards->push_back(cardToAdd);
 }
 
 void Deck::shuffleCards()
 {
-  std::random_shuffle (_cards.begin(), _cards.end());
+  std::random_shuffle(cards->begin(), cards->end());
 }
 
 Card Deck::moveCard(Deck &ToThisDeck, Card cardToMove)
 {
-  for(uint i = 0; i < _cards.size(); ++i)
+  uint cardIndex = findCard(cardToMove);
+  ToThisDeck.cards->push_back(this->cards->at(cardIndex));
+  removeCard(cardToMove);
+  return cardToMove;
+}
+
+std::unique_ptr<Card> Deck::removeCard(Card cardToRemove)
+{
+  std::unique_ptr<Card> removedCard;
+  if(cards->size() <= 0)
   {
-    if(_cards.at(i) == cardToMove)
-    {
-      ToThisDeck._cards.push_back(_cards.at(i));
-      _cards.erase(_cards.begin() + i);
-    }
+    std::cout << "Váš balíček karet je prázdný";
   }
-  return cardToMove;
+  else
+  {
+    uint cardIndex = findCard(cardToRemove);
+    removedCard = std::make_unique<Card> (this->cards->at(cardIndex));
+    this->cards->erase(this->cards->begin() + cardIndex);
+  }
+  return removedCard;

```

Comment:
zamiast tworzyć removedCard i go zwracać to lepiej zwrócić std::move(poprzedni_uniqe) przez co od razu mamy move semantic i zwolnienie poprzedniego unique_ptr.

Reply:
Czaję o co chodzi z std::movem, ale czy ten poprzedni pointer ma być pointerem na ostatnią usuniętą kartę? Jeśli tak to mam go deklarować globalnie i przy każdym find zwalniać i uzupełniać?

Reply:
@mnurzyns

Reply:
Jaki poprzedni pointer? Nic globalnego też nie musisz mieć.
1. Przyjmujesz cardToRemove (zmień to na referencję).
2. Znajdujesz jego indeks
3. Mając indeks robisz std::move tej karty w returnie

Aby to ładnie napisać to w ifie zwróc nullptr. Dodatkowo teraz widzę problem z tym co jeżeli nie znajdzie Ci danej karty? findCard nie odnajdzie tej karty i już będzie problem z tym co usunąć.